app/src/main/java/com/quran/labs/androidquran/widgets/HighlightingImageView.java

```diff
@@ -130,17 +152,178 @@ public void setNightMode(boolean isNightMode, int textBrightness) {
     adjustNightMode();
   }
 
-  public void highlightAyah(int sura, int ayah, HighlightType type) {
+  private void highlightFloatableAyah(Set<String> highlights, String currentSurahAyah) {
+    final Map<String, List<AyahBounds>> coordinatesData = ayahCoordinates.getAyahCoordinates();
+
+    String previousSurahAyah = currentSurahAyah;
+    for(String surahAyahs: highlights) {
+      previousSurahAyah = surahAyahs;
+    }
+    List<AyahBounds> startingBounds;
+    int arrowIndex = previousSurahAyah.indexOf("->");
+    if(arrowIndex != -1) {
+      // The ayah changed during animating
+      startingBounds = (List<AyahBounds>)animator.getAnimatedValue();
+      animator.cancel();
+      previousSurahAyah = previousSurahAyah.substring(arrowIndex + 2);
+    } else {
+      startingBounds = coordinatesData.get(previousSurahAyah);
+    }
+
+    final String ayahTransition = previousSurahAyah+"->"+currentSurahAyah;
+
+    // yes we make copies, because normalizing the bounds will change them
+    List<AyahBounds> previousAyahBoundsList = new ArrayList<>(startingBounds);
+    List<AyahBounds> currentAyahBoundsList = new ArrayList<>(coordinatesData.get(currentSurahAyah));
+
+    highlights.clear();
+    highlights.add(ayahTransition);
+
+    // TODO: reason whether this TypeEvaluator should be anonymous or not
+    animator = ValueAnimator.ofObject(new TypeEvaluator() {
+
+      private void normalizeAyahBoundsList(List<AyahBounds> start, List<AyahBounds> end) {
+        // this function takes two unequal length lists and normalizes them
+
+        int startSize = start.size();
+        int endSize = end.size();
+        int minSize = Math.min(startSize, endSize);
+        int maxSize = Math.max(startSize, endSize);
+        List<AyahBounds> minList = startSize < endSize? start : end;
+        int diff = maxSize - minSize;
+
+        RectF rectToBeDivided = minList.get(minSize-1).getBounds();
+        float oLeft = rectToBeDivided.left;
+        float oRight = rectToBeDivided.right;
+        float oTop = rectToBeDivided.top;
+        float oBottom = rectToBeDivided.bottom;
```

Comment:
what's the `o` here for? can we expand it?

Reply:
have expanded it to original. It refers to the original rectangle that has to be divided